src/cosl/coordinated_workers/interface.py

```diff
@@ -403,6 +406,29 @@ def get_address_from_role(self, role: str) -> Optional[str]:
             return address_set.pop()
         return None
 
+    def _has_worker_published(self, relation: ops.Relation) -> bool:
+        """Verify that the worker has published its data.
+
+        - unit address is published
+        - roles are published
+        """
+        if not relation.app:
+            return False
+        if len(relation.units) == 0:
+            return False
+        for worker_unit in relation.units:
+            try:
+                ClusterRequirerUnitData.load(relation.data[worker_unit])
+                if self._charm.unit.is_leader():
+                    ClusterRequirerAppData.load(relation.data[relation.app])
```

Comment:
and actually we shouldn't be doing this either inside the for loop. it's enough to check app data once.